### ISSUE TITLE
support bootstrapping legacy mvapich v1 builds in Flux

### ISSUE THREAD

**Issue by dongahn** [flux-framework/flux-core#221]
Flux should support running codes compiled against older mvapich v1 MPI, which does not use PMI for bootstrap.  Instead it connects to `MPIRUN_HOST:MPIRUN_PORT` to exchange personality info.  The LC modified mvapich v1 uses an improved collective-based protocol for bootstrap called PMGR.  Ideally both the upstream and PMGR mvapich v1 should work with Flux, as it does with Slurm, without recompilation of MPI libraries or applications.

**Comment by dongahn** [comment]
I misspoke. I am taking a look at the old code from the NRGM repo, and I actually wrote pmgr collective on top of kvs directly instead of our PMI (though the binding is OLD -- cmb_kvs_put :-). This should have a slight performance advantage. I will probably want to tease apart the PMGR code base a bit only to get the parts Flux needs.

Two questions:
1. What KVS directory PMI currently puts/gets the information for the job? If PMI somehow gets passed the information on lwj.<id> and puts/gets the key-value pair inside the target LWJ (sorry for the old terminology), I would like to replicate that to PMGR code. It currents puts the info in a different directory than lwj. 

2. Named barrier. PMGR uses Flux's name barrier to synchronize among distributed processes to make sure all of the processes put their identity and flush/commit. I vaguely remember the  name must be unique for each barrier instance. Is this restriction still there?

**Comment by garlick** [comment]
The current Flux PMI code uses the prefix pmi.lwjid to store keys in the KVS.  That was an ad hoc choice a while back and should probably be changed as part of a comprehensive "program context" design.   If you were to port pmgr on top of pmi, then the pmgr code wouldn't have to track those changes.  What is the performance reason not to do that?

Internally, the PMI library maps `PMI_KVS_Commit()` to nothing, and `PMI_Barrier()` to `kvs_fence()`.  That fit the mvapich2 and openmpi usage that I was tracking, and as I recall I read that this was ok to assume `PMI_Barrier()` was not going to be used as a general purpose barrier in an mpich PMI implementors note somewhere in the source tree.  I did not implement PMIv2 calls though that was/is planned.

**Comment by garlick** [comment]
@dongahn - should the title of this issue by "support bootstrapping mvapich v1 in Flux?  What are the mechanics of that - do we provide a replacement pmgr library like we provide pmi?  Sorry if I'm forgetting how the pieces fit together.

**Comment by grondo** [comment]
@garlick, there isn't a pmgr library with mvpaich per se. In the
srun/mpirun case, the launcher exposes a port via `MPIRUN_HOST`
`MPIRUN_PORT` environment variables, and in different versions of the
bootstrap protocol all or some ranks connect back to this port, which
provides a service that speaks the pmgr protocol. Previous to pmgr, there
was an even more rudimentary and less scalable protocol, but mechanics were
similar.

The code that implements the protocol in srun is found here:


https://github.com/chaos/slurm/blob/slurm-2.3.3.chaos/src/plugins/mpi/mvapich/mvapich.c

**Comment by dongahn** [comment]
@garlick I will probably need to take a look at MVAPICH  source (probably inside mpi init) as well to see how to enable the code that uses pmgr to bootstrap the mpi fabric. My guess is the same mvapich that works today in our Linux cluster will work with essentially no modification if it links in our pmgr port. 

I don't want to call this issue MVAPICH support because pmgr can support other MPIs and even tools.

In fact I will port COBO as a thin layer on top of PMGR which then can be the basis for a part of LauchMON and other runtime tools.

**Comment by dongahn** [comment]
By the way, in terms of porting PMGR on top of PMI vs KVS, either should be a reasonable approach.

I slightly prefer the former because I already have a proof of concept using this approach,  this seems to have a slight performance advantage, and I sort of see pmgr at the same level of as PMI.

I am open to other thoughs though.

Best
Dong

**Comment by garlick** [comment]
Maybe there are a two different issues here.  One is for Flux to provide the PMGR _service_ that early mvapich/mpich needs to bootstrap, as implemented in Slurm by the code @grondo referenced.  That might be needed to support "off the shelf" MPI's derived from older mpich, and it would be useful to us now because the older mvapich is LC's current default.  Maybe we could do this by setting `MPIRUN_HOST=127.0.0.1` and `MPIRUN_PORT` to a port listened to by a comms module that speaks the PMGR protocol and use the KVS underneath to exchange data securely and scalably.

Then there is tool bootstrapping, where @dongahn is the expert and I tend to get lost :-(  We have LaunchMon covered in issue #177.  Is that where we should be talking about what Flux needs to provide in this area?   From my naive point of view, it would seem like we should be able to graft LaunchMon directly onto Flux and leave PMGR/COBO out of it but I probably am basing that on an incomplete understanding.

**Comment by dongahn** [comment]
Agreed that we should use the discussion for LaunchMON using our placeholder at #177.

Let's discuss a bit more about the MPIRUN_HOST/MPIRUN_PORT approach, perhaps we are talking about a similar mechanism. 

BTW, I don't think providing the PMGR protocol like what slurm does is a scalable approach. In this scheme, the execution service will launch processes to the remote nodes. Then, each process will call MPI_Init and within that, MPI will bootstrap pmgr collectives (rudimentary cast, reduce, gather/scatter) to help setting up the MPI fabric scalable. For scalability, we probably don't want each MPI process connect to MPIRUN_PORT.

Instead, it would be better if the remote processes exchange their personality information through KVS. My code does this by connecting to the local cmbd with flux_open and uses kvs module-interface library to exchange the personality information. 

This approach is described in my old slides deck (slide #13) that I used for CEA meeting Sep/2013. I will send this to the folks. (Headed to the group meeting...)

**Comment by garlick** [comment]
My impression from the code referenced above and from briefly talking with @grondo this morning was that one could  trick the PMGR code in mvapich to connect to a local port (on each compute node) that was perhaps managed by a Flux comms module, and then build a distributed service on the KVS to perform the exchange/reduction, and give back the personality info over the same local port.  The hard part would be implementing the PMGR protocol, but hopefully one of these libraries already has that done in a manner that we could borrow?   If nothing else, maybe the slurm plugin could be adapted.

**Comment by dongahn** [comment]
OK. I talked to Adam:

1) I will have to change the title for this issue:  supporting legacy mvapich1 builds (written on top of PMGR collective).

2) For mvapich1, @grondo is correct: it was locally patched to include PMGR source files. So redirecting its PMGR requirement to Flux-provided PMGR library is not going to do the work. For mvapich2 builds, they call PMI via a shared library, so we can us that trick.  

3) We may be able to satisfy most of what PMGR needs by exporting a set of environment variables:
     SLURM_NODEID, SLURM_LOCALID, SLURM_STEP_NODELIST, SLURM_STEP_TASKS_PER_NODE ...

This seems a good news because we won't need a compat module for this. I guess we can simply export these environment variables through wreckrun modules. 

4) But when MPI calls MPI_Abort, mvapich1 (via pmgr_abort) connects to the starter process and speaks PMGR's abort protocol. To satisfy this, it seems we will need a compat module...

I think this is doable, but this actually reminds me of a question. Is it a good practice (or easy to do) for a comms module to open a TCP socket and register a callback to be invoked on a connect?

**Comment by garlick** [comment]
Now it makes sense.

For the record, it should be fine to implement a TCP service in a comms module.  (See modules/api for one that does this with a UNIX domain socket, pretty much the same thing)

I opened #223 to remind us to clean all those SLURM_ variables out of the environment to avoid confusing the MPI runtimes, since what they are inheriting from the original Slurm allocation is potentially a mismatch.

I guess to launch mvapich v1 we might want to add some sort of compat option?  (I don't think we want to be setting those SLURM_ variables by default).

**Comment by dongahn** [comment]
@garlick makes sense. 

Perhaps, the user interface for this option is provided as a wreckrun option (e.g., --bootstrap-compat or something) and have an argument to the option to support a range of lagacy MPI binaries? (e.g., --bootstrap-compat=pmgr).

**Comment by grondo** [comment]
@dongahn -- I was hoping to get away from select-mpi-with-runtime-option style launch like SLURM does with `srun --mpi=foo`. However, for practical purposes this might not be possible because various MPICH-based MPIs have overloaded the use of these environment variables heavily :frowning:
This leads me to think that, as in SLURM, we can always support PMI (via libpmi.so), but we may have to support a `--mpi` mpi selection option for program launch so that the correct service is listening on `MPIRUN_HOST` & `MPIRUN_PORT`

**Comment by dongahn** [comment]
--mpi= option works for me. Over time, MPIs and runtime tools converge on known bootstrapping interfaces (via  a shared library), so we may have a chance to clean this up.

**Comment by dongahn** [comment]
I looked at the wrexec.c and wrexecd.c, and identified the code location to export those environment variables (with @grondo 's help). 

One issue is that, because neither the wrexec subsystem nor KVS retains the information about the hostname lists (e.g., the one that exposes as SLURM_NODELIST and the other SLURM_STEP_NODELIST), it looks like I will have to generate the info. 

Since this seems generally useful, I am inclined to create a reduction function that takes the numbers used in the hostnames and computes the condensed host list form in parallel (e.g., cab[2-3, 4]).

I've looked at some codes (e.g., live.c and reduce.h), and I think I understand most of the logic behind Flux's reduction, but I have several questions before coding this up. 

1. I want to use nodeset_t as the base type to reduce the host numbers. However, the raw nodeset_t representation probably won't work as a message payload, does it?  If not, my second best is to wrap it with a JSON object, but this type doesn't map to one primitive data type so I'm currently thinking to covert it to the string. Does this sound reasonable? If there are other ways to make nodeset_t work better with reduction/json,  I'd like to learn more about it. 

2. This reduction will be more or less deterministic -- similar to collective reduction like MPI_reduce. Each rank will contribute one item, and all the items will always be reduced to down to one item. As such, I don't want to use Timed reduction (FLUX_RED_TIMEDFLUSH), which will add unnecessary latency. Is FLUX_RED_HWMFLUSH  a way to go? (I couldn't follow exactly what the comment says; probably me :-).

3. Finally, I actually lied. This is more like MPI_allreduce. So the what's reduced to the root will have to be broadcast to the rest of the ranks. How does the cmbd root determine all items are reduced, so that it can broadcast the reduced item to the rest of the system.

Apology -- I would have formulated these questions better if I had spent a bit more time; but today's my son's birthday party so I need to go now. We can also chat during our meeting as well, though.

**Comment by springme** [comment]
I hope that you made it to the birthday party in time!



Sent with Good (www.good.com)


-----Original Message-----

**Comment by garlick** [comment]
Hi dong, probably nodeset is inappropriate here as it doesn't permit duplicates.  

I can help with reduction but maybe flux should already know _its_ host-rank mapping from its enclosing instance and should provide subset mapping to programs it launches, ie without need to compute on the fly?

**Comment by grondo** [comment]
Programs launched within an instance will necessarily know the list of ranks on which their tasks are being launched. If we reinstated the rank-to-hostname mapping in flux, then the program hostname list would be easy to reconstruct. We'd discussed briefly a generic lookup service in a flux instance, but we decided it wasn't necessary or tabled the issue for some reason (and I can no longer find the discussion). It seems like it would be nice to easily map program *task* ids to ranks, and ranks to hostnames or other data, and to have a standard way to store and retrieve that data in an instance.

I think Dong was proposing to establish the rank-to-hostname mapping in the startup of the `exec` module, not once per program launch. Reduction does seem like overkill here, but I think he's probably looking for an excuse to play with the interface. :smile:

**Comment by garlick** [comment]
That would be good then.  Those interfaces could use some review especially from @dongahn.

**Comment by dongahn** [comment]
@garlick 

Thanks. I looked at nodeset_t because I thought a generalized nodeset_t could be a good data type that can support a parallel reduction over two sets of integer lists. I guess this operation would be very common. For example STAT uses a similar algorithm for fast reduction of an MPI rank set per each call path.

I thought that I could use this for reduction by converting the RHS nodeset_t into a string and add that string to the LHS nodeset_t object.

Once we had a host names list in KVS ordered by cmbd rank. That structure isn't available.

So as part of JSC, I simply use a cmbd rank (after itoa) as the hostname...

**Comment by dongahn** [comment]
@grondo yes that kind of service is what we will need. 

And yes I am proposing to generate the mapping wrexec once and use it for SLURM_NODELIST. Then for each session (a program) we need to have a similar method to expose SLURM_STEP_NODELIST.

Finally yes (you are reading my mind :-)  I am taking this opportunity to examine reduction and some other collective communication patterns within Flux. 

If these Flux interfaces map well to the patterns like pmgr_gather reduce etc, reimplementing the PMGR interface directly off of them may be straightforward. 

I think I understand the general mechanism of Flux's reduction but I thought I wanted to generate some discussions on them.

In particular I wondered if one could raise its reduction abstraction level  a bit higher so that it's users wont even have to call flux event-message sends on sink and also not explicitly register a raw event handler. Briefly looking at the code I thought flux_red_t can be extended to support these.

But this was just from a brief inspection as such please take this w/ a grain of salt.

**Comment by garlick** [comment]
Yes I had the same thought about extending the reductions to cover message handling.  

Keep in mind program ranks may not have unique hostname so reducing numerical suffux using nodeset that doesn't allow dups and doesn't preserve order could be problem.  Hostlist_t may be more appropriate.

**Comment by dongahn** [comment]
That'd be great if the reduction can hide the message handling. Most of patterns I've seen hide it. 

Yes the nodeset_t won't work if the instance goes across different clusters. 

Come to think of this, making those reducible data types as generic as possible so that we could use them for various reduction patterns seems like something to think about.

**Comment by dongahn** [comment]
It looks to me the current libflux reduction  support can be used for this purpose. `SLURM_NODELIST` and `SLURM_STEP_NODELIST` are the ones that can be scalably exported using distributed reduction.

BTW, this was just a driver to understand the current reduce support, and my current conclusion after prototyping with `flux_red_t` is that the reduction can be further extended into a module to provide a bit better isolation. As is, it seems difficult to create a logically isolated reduction operation without having to worry about other overlapping reduction operations (of the same type.)

In this particular case of having to reduce hostnames, it is ideal if the same hostname reduction operations performed by wrexec and wrexecd sessions are all logically separated. Overall, if we were to generalize such extended reduction support through a module, isolation would be not only good for programmability but also security. We probably don't want one program owned by user A maliciously to try to have its reduction input data merged into an overlapping reduction operation issued by another program owned by user B. 

I think we can solve the isolation problem with 'named' reduction approach similar to the current named barrier. As mentioned in #251, I am creating a new idea issue to discuss general named reduction support.

**Comment by garlick** [comment]
This discussion is pretty old and it looks like we are finally migrating away from the older, pre-PMI mvapich1 software.  Let's close this for now and reopen if we need to.
